File: hw4_xiangl2/alarm.cs
using System;

class alarm {
    static void Main() {
        string line;
        while((line = Console.ReadLine()) != null && line != ""){
            if(line.Length < 11){
                line = Console.ReadLine();
            }
            int pushes = 0;
            time startTime = new time(-1, -1, -1, "MM");
            time endTime = new time(-1, -1, -1, "MM");

            startTime.setHour(line[0] - '0');
            startTime.setM1(line[2] - '0');
            startTime.setM2(line[3] - '0');
            startTime.setAp(line[4].ToString() + "m");

            endTime.setHour(line[7] - '0');
            endTime.setM1(line[9] - '0');
            endTime.setM2(line[10] - '0');
            endTime.setAp(line[11].ToString() + "m");

            int hDif = Math.Abs(startTime.getHour() - endTime.getHour());
            pushes += Math.Min(hDif, 12 - hDif);

            int m1Dif = Math.Abs(startTime.getM1() - endTime.getM1());
            pushes += Math.Min(m1Dif, 6 - m1Dif);

            int m2Dif = Math.Abs(startTime.getM2() - endTime.getM2());
            pushes += Math.Min(m2Dif, 10 - m2Dif);

            if(!string.Equals(startTime.getAp(), endTime.getAp())){
                pushes += 1;
            }
            
            if(pushes == 1){
                Console.Write("Going from " + startTime.display() + " to "  + endTime.display() + " requires "  + pushes +  " push\n");
            }else {
                Console.Write("Going from " + startTime.display() + " to "  + endTime.display() + " requires "  + pushes +  " pushes\n");
            }

        }

    }

    public class time {
        public int hour, m1, m2;
        public string ap;

        // time constructor
        public time(int hour, int m1, int m2, string ap){
            this.hour = hour;
            this.m1 = m1;
            this.m2 = m2;
            this.ap = ap;
        }

        public int getHour(){
            return hour;
        }
        public void setHour(int n){
            this.hour = n;
        }
        public int getM1(){
            return m1;
        }
        public void setM1(int n) {
            this.m1 = n;
        }
        public int getM2(){
            return m2;
        }
        public void setM2(int n){
            this.m2 = n;
        }
        
        public string getAp(){
            return ap;
        }

        public void setAp(string a){
            this.ap = a;
        }

        public string display(){
            return (this.getHour() + ":" + this.getM1() + this.getM2() + this.getAp());
        }
    }
}

File: HW6S19/hw6/alarm.py
import sys
import math

# create an obj of time
class time:
    def __init__(self, hour, m1, m2, ap):
        self.hour = hour
        self.m1 = m1
        self.m2 = m2
        self.ap = ap
    
    def display(self):
        s = str(self.hour) + ":" + str(self.m1) + str(self.m2) + str(self.ap) + "m"
        return s

def seperate(s):
    startTime.hour = int(s[0])
    startTime.m1 = int(s[2])
    startTime.m2 = int(s[3])
    startTime.ap = s[4]

    endTime.hour = int(s[7])
    endTime.m1 = int(s[9])
    endTime.m2 = int(s[10])
    endTime.ap = s[11]


startTime = time(-1, -1, -1, "P")
endTime = time(-1, -1, -1, "P")


#read line from the input which has a form of
# 0:00AM 0:00AM
while True:
    line = sys.stdin.readline() 
    if len(line) == 0: break
    if(len(line) < 11): line = sys.stdin.readline()
    
    push = 0
    seperate(line)

    hDif = math.fabs(startTime.hour - endTime.hour)
    push += min(hDif, (12 - hDif))

    m1Dif = math.fabs(startTime.m1 - endTime.m1)
    push += min(m1Dif, (6 - m1Dif))

    m2Dif = math.fabs(startTime.m2 - endTime.m2)
    push += min(m2Dif, (10 - m2Dif))

    if(startTime.ap != endTime.ap):
        push += 1
    
    if(push == 1):
            print("Going from %s to %s requires %s push" %(startTime.display(), endTime.display(), int(push)))
    else:
        print("Going from %s to %s requires %s pushes" %(startTime.display(), endTime.display(), int(push)))

